core/bentley/src/Logger.ts

```diff
@@ -58,101 +52,58 @@ export interface LoggerLevelsConfig {
  * @public
  */
 export class Logger {
-  private static _logError: LogFunction | undefined;
-  private static _logWarning: LogFunction | undefined;
-  private static _logInfo: LogFunction | undefined;
-  private static _logTrace: LogFunction | undefined;
-  private static _logIntercept: LogIntercept | undefined;
+  protected static _logError: LogFunction | undefined;
+  protected static _logWarning: LogFunction | undefined;
+  protected static _logInfo: LogFunction | undefined;
+  protected static _logTrace: LogFunction | undefined;
   private static _categoryFilter: Map<string, LogLevel> = new Map<string, LogLevel>();
   private static _minLevel: LogLevel | undefined = undefined;
   private static _logExceptionCallstacks = false;
-  private static _makeMetaDataEvent: BeEvent<(...arg: any[]) => void> = new BeEvent<(...arg: any[]) => void>();
+  /** @internal */
+  public static staticMetaData: ExceptionMetaData;
```

Comment:
I can see how this is simpler/cleaner than using `BeEvent`/callbacks/etc. (and the whole "...MetaDataFromSources" naming was terrible), but I'm worried this is a bit harder to use now than it was.

For example, we have a number of places in our agents where we want to add some metadata to all future log messages.  In a few cases we do that multiple times (e.g., v2-checkpoint-agent adds imodelId/changesetId to all logs at startup, but has to actually query the changeset index and call `registerMetaDataSource` again to also add _that_ to all future logs).

With just the one `staticMetaData` property here adding that second set of properties gets much harder, since now I have to detect whether the existing `staticMetaData` is an object or function and wrap it accordingly.

Maybe this would be easier if we said `staticMetaData` just always had to be an object?

Reply:
if there are multiple sources, they should each be logged separately, not combined into one object. I made staticMetaData be a Map<string,LoggingMetaData>. See what you think.

Reply:
i could add the "label" in the log message. would that be helpful?

Reply:
This seems fine for logging to the console, but I think I'm still going to have to combine this all into one flat object when I log to something like SEQ  - otherwise it's going to be a **lot** harder to query logs from deployed backends/agents, and we'd break just about every saved log query we have.

Reply:
I don’t understand. How does this complicate searching logs? Combining separate static objects into one is absolutely the wrong thing to do. It intermingles the properties, hiding their ownership, and even potentially losing values if the same name exists multiple places. I cannot understand any reason to do that.

Or, are you just worried about backwards compatibility for that terrible idea?

Reply:
SEQ logs are structured - each log message we write is a JSON object, where the actual "message" is just a string property.  Other primitive properties on that object can potentially be indexed and are easily searchable - so this is where we really need things like ActivityId, iModelId, etc. to be.

Objects can also be properties of the message object, but I don't think _their_ properties are usable in SEQ signals - which means they cannot be indexed and will be much harder to search for.  And yeah, backwards compatibility is also a concern here - there are many existing queries and signals that rely on events having an "ActivityId" or "iModelId" property.  Even if we aren't concerned with breaking those queries, those properties are standardized across many services, so any change to them here will make correlating logs across services a nightmare.

You're absolutely right that there are risks in merging all these objects together, but without being able to add global _primitive_ properties directly to every log object, our SEQ logs will be nearly useless.

Reply:
Can we serialize them into an object with member objects, or does seq only index top level entries?

Reply:
i merged them all together. Please review.

Reply:
LGTM

Reply:
@wgoehrig i switched the order so the per-call-metadata takes precedence. Please re-review.